### PR TITLE
Separate transport related features from Swarm<T>

### PR DESCRIPTION
**Pull request by limebell**
I moved NetMQ and STUN related stuffs to new class called `NetMQTransport` which inherits `ITransport`.
This leads to some changes, which are:
- `KademliaProtocol` now takes `ITransport` as its first argument in constructor, instead of ISwarm.
- Removed `Swarm<T>.Protocol` internal property since it's moved to `NetMQTransport`.
- `Protocol` benchmark removed.
- `Swarm<T>` became not to implement `ISwarm`.
- `TestSwarm` used in some tests renamed to `TestProtocol`, which implements `ITransport` instead of `ISwarm`.

### CODE REVIEW

File: Libplanet.Tests/Net/Protocols/TestTransport.cs [hunk 1]
```diff
@@ -68,47 +67,63 @@ public class TestSwarm : ISwarm
...
-            await ProcessRuntime(swarmCancellationTokenSource.Token);
+            _logger.Debug("Starting transport of {Peer}.", AsPeer);
+            _swarmCancellationTokenSource = new CancellationTokenSource();
+            await Task.Delay(10, cancellationToken);
         }
 
-        public Task BootstrapAsync(
-            IEnumerable<BoundPeer> peers,
+        public async Task RunAsync(
+            CancellationToken cancellationToken = default(CancellationToken))
```
Comment (dahlia) [inline Libplanet.Tests/Net/Protocols/TestTransport.cs]: Why is this `cancellationToken` not used?

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 2]
```diff
@@ -2063,6 +2063,9 @@ public async Task PreloadSparseStateWithTrustedPeers()
                     }),
                     trustedStateValidators: trustedPeers);
 
+                // Await 1 second to make sure all progresses is reported.
+                await Task.Delay(1000);
```
Comment (dahlia) [inline Libplanet.Tests/Net/SwarmTest.cs]: How about using [`EventWaitHandle`][1] to wait until all progresses are reported?

[1]: https://docs.microsoft.com/en-us/dotnet/api/system.threading.eventwaithandle?view=netstandard-2.0

File: Libplanet.Tests/Net/Protocols/TestTransport.cs [hunk 3]
```diff
@@ -68,47 +67,69 @@ public class TestSwarm : ISwarm
...
-        public async void Start()
+        public async Task StartAsync(
+            CancellationToken cancellationToken = default(CancellationToken))
         {
-            _logger.Debug("Starting swarm {Peer}.", AsPeer);
-            swarmCancellationTokenSource = new CancellationTokenSource();
-            await ProcessRuntime(swarmCancellationTokenSource.Token);
+            _logger.Debug("Starting transport of {Peer}.", AsPeer);
+            _swarmCancellationTokenSource = new CancellationTokenSource();
+            await Task.Delay(10, cancellationToken);
```
Comment (dahlia) [inline Libplanet.Tests/Net/Protocols/TestTransport.cs]: Why is this delay needed?

File: Libplanet.Tests/Net/Protocols/TestTransport.cs [hunk 4]
```diff
@@ -68,47 +67,69 @@ public class TestSwarm : ISwarm
...
         }
 
-        public Task BootstrapAsync(
-            IEnumerable<BoundPeer> peers,
+        public async Task RunAsync(
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var token = cancellationToken == default(CancellationToken)
```
Comment (dahlia) [inline Libplanet.Tests/Net/Protocols/TestTransport.cs]:
```suggestion
            CancellationToken token = cancellationToken.Equals(CancellationToken.None)
```

File: Libplanet.Tests/Net/Protocols/TestTransport.cs [hunk 5]
```diff
@@ -68,47 +67,69 @@ public class TestSwarm : ISwarm
...
-                    peers.ToImmutableList(),
-                    pingSeedTimeout,
-                    findPeerTimeout,
-                    Kademlia.MaxDepth,
-                    cancellationToken);
-            }
-
-            return DoBootstrapAsync();
+            await Protocol.BootstrapAsync(
+                bootstrapPeers.ToImmutableList(),
```
Comment (dahlia) [inline Libplanet.Tests/Net/Protocols/TestTransport.cs]: Isn't this okay without `.ToImmutableList()`?

```suggestion
                bootstrapPeers,
```

File: Libplanet.Tests/Net/Protocols/TestTransport.cs (hunk 3 above)
Comment (limebell) [reply Libplanet.Tests/Net/Protocols/TestTransport.cs]: I added it because `ITransport` needs `RunAsync()` to be `async`, but `TestTransport`'s `RunAsync()` need not to be `async`.

File: Libplanet/Net/DifferentProtocolVersionEventArgs.cs [hunk 6]
```diff
@@ -2,7 +2,7 @@ namespace Libplanet.Net
 {
     /// <summary>
     /// The event data that provides a value when
-    /// <see cref="Swarm{T}.DifferentVersionPeerEncountered" /> is occurred.
+    /// <see cref="NetMQTransport.DifferentVersionPeerEncountered" /> is occurred.
```
Comment (longfin) [inline Libplanet/Net/DifferentProtocolVersionEventArgs.cs]: I guess `DiffrentVersionPeerEncountered` isn't related to specific transport like `NetMQTransport`.

File: Libplanet/Net/NetMQTransport.cs [hunk 7]
```diff
@@ -0,0 +1,1009 @@
...
+        private readonly ILogger _logger;
+
+        private NetMQQueue<Message> _replyQueue;
+        private NetMQQueue<(Address?, Message)> _broadcastQueue;
+
+        private RouterSocket _router;
+        private NetMQPoller _poller;
+
+        private int? _listenPort;
+        private TurnClient _turnClient;
```
Comment (longfin) [inline Libplanet/Net/NetMQTransport.cs]: TURN related features also don't seem related to NetMQTransport.

File: Libplanet/Net/Swarm.cs [hunk 8]
```diff
@@ -218,31 +135,18 @@ static Swarm()
...
-        /// <summary>
-        /// The <see cref="EventHandler" /> triggered when the different version of
-        /// <see cref="Peer" /> is discovered.
-        /// </summary>
-        public event EventHandler<DifferentProtocolVersionEventArgs>
-            DifferentVersionPeerEncountered;
+        public bool Running => _transport is NetMQTransport p && p.Running;
 
-        public DnsEndPoint EndPoint { get; private set; }
+        public DnsEndPoint EndPoint => (_transport as NetMQTransport)?.EndPoint;
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]: How about adding `EndPoint` as property to `ITransport`?

File: Libplanet/Net/Swarm.cs [hunk 9]
```diff
@@ -407,78 +270,12 @@ public void Dispose()
...
-                _listenPort = _router.BindRandomPort("tcp://*");
-            }
-            else
-            {
-                _router.Bind($"tcp://*:{_listenPort}");
-            }
-
-            _logger.Information($"Listen on {_listenPort}");
-
+            List<Task> tasks = new List<Task>();
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]:
```suggestion
            var tasks = new List<Task>();
```

File: Libplanet/Net/Swarm.cs [hunk 10]
```diff
@@ -2076,7 +1617,7 @@ private void TransferBlocks(GetBlocks getData)
             int total = hashes.Count;
             const string logMsg =
                 "Fetching a block #{Index}/{Total} ({Hash}) to include to " +
-                "a reply to {Identity}...";
+                "a reply 1to {Identity}...";
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]:
```suggestion
                "a reply to {Identity}...";
```

File: Libplanet/Net/Swarm.cs [hunk 11]
```diff
@@ -305,15 +189,13 @@ private set
...
         /// property becomes <c>true</c>.</returns>
-        public Task WaitForRunningAsync() => _runningEvent.Task;
+        public Task WaitForRunningAsync() => (_transport as NetMQTransport)?.WaitForRunningAsync();
 
         public void Dispose()
         {
-            _runtimeCancellationTokenSource.Cancel();
-
-            _runtimeProcessor.Wait();
+            (_transport as NetMQTransport)?.Dispose();
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]: How about extending `ITransport` as  `IDisposable`?

File: Libplanet/Net/Swarm.cs [hunk 12]
```diff
@@ -218,31 +135,18 @@ static Swarm()
...
-            ? new Peer(
-                _privateKey.PublicKey,
-                _appProtocolVersion,
-                _publicIPAddress)
-            : new BoundPeer(
-                _privateKey.PublicKey,
-                EndPoint,
-                _appProtocolVersion,
-                _publicIPAddress);
+        public Peer AsPeer => (_transport as NetMQTransport)?.AsPeer;
```
Comment (longfin) [inline Libplanet/Net/Swarm.cs]: How about adding `AsPeer` property to `ITransport`?

File: Libplanet/Net/DifferentProtocolVersionEventArgs.cs (hunk 6 above)
Comment (limebell) [reply Libplanet/Net/DifferentProtocolVersionEventArgs.cs]: You are right. I left `FIXME` in protocol detection part in `NetMQTransport` that it should be in `Swarm<T>`.

File: Libplanet/Net/NetMQTransport.cs [hunk 13]
```diff
@@ -0,0 +1,1009 @@
...
+                () =>
+                {
+                    // Ignore NetMQ related exceptions during NetMQPoller.Run() to stabilize
+                    // tests.
+                    try
+                    {
+                        _poller.Run();
+                    }
+                    catch (TerminatingException)
+                    {
```
Comment (moreal) [inline Libplanet/Net/NetMQTransport.cs]: It seems some lines needed more. see also [Swarm.cs#L514](https://github.com/planetarium/libplanet/blob/master/Libplanet/Net/Swarm.cs#L514).

File: Libplanet/Net/NetMQTransport.cs (hunk 7 above)
Comment (limebell) [reply Libplanet/Net/NetMQTransport.cs]: I guess it will be great to be done in another PR. I'll left an issue, and a FIXME comment either.

File: Libplanet/Net/NetMQTransport.cs [hunk 14]
```diff
@@ -0,0 +1,1009 @@
...
+            }
+            catch (InvalidMessageException ex)
+            {
+                _logger.Error(ex, $"Could not parse NetMQMessage properly; ignore: {ex}");
+            }
+            catch (Exception ex)
+            {
+                _logger.Error(
+                    ex,
+                    $"An unexpected exception occurred during ReceiveMessage(): {ex}"
```
Comment (moreal) [inline Libplanet/Net/NetMQTransport.cs]: How about this?

```suggestion
                    $"An unexpected exception occurred during {nameof(ReceiveMessage)}(): {ex}"
```

File: Libplanet/Net/NetMQTransport.cs (hunk 13 above)
Comment (limebell) [reply Libplanet/Net/NetMQTransport.cs]: I added it :)

File: Libplanet/Net/NetMQTransport.cs [hunk 15]
```diff
@@ -0,0 +1,1011 @@
...
+
+        /// <summary>
+        /// The <see cref="EventHandler" /> triggered when the different version of
+        /// <see cref="Peer" /> is discovered.
+        /// </summary>
+        private event EventHandler<DifferentProtocolVersionEventArgs>
+            DifferentVersionPeerEncountered;
+
+        /// <summary>
+        /// The <see cref="EventHandler" /> triggered when a <see cref="Message"/> is
```
Comment (moreal) [inline Libplanet/Net/NetMQTransport.cs]: It's just question, is there some rules to add space before the end?

```
        /// The <see cref="EventHandler" /> triggered when a <see cref="Message"/> is
                                        ^ here                                 ^ but not this 🤔 
```

Comment (limebell) [reply Libplanet/Net/NetMQTransport.cs]: I guess not...

Comment (dahlia) [reply Libplanet/Net/NetMQTransport.cs]: Both are valid as it's XML, but I prefer no space, because we usually have no enough horizontal space.

Comment (limebell) [reply Libplanet/Net/NetMQTransport.cs]: I found that existing xml documents also have this problem. This will be great to be dealt in new PR .
